test-documentation: Create Python version of test-documentation with check for virtual environment.

Shashantao opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
Fixes: #17929 

**Testing plan:** <!-- How have you tested? -->
We transpiled `test-documentation` to a Python script named `test-documentation-py-shash`. We compared the outputs of `test-documentation` and `test-documentation-py-shash` with the command line arguments `--skip-check-links`, `--skip-external-links`, `-L ERROR`, and `-L DEBUG`, respectively; we verified that the outputs of both scripts were equivalent with the following command for example:
`diff <(./tools/test-documentation-py-shash -L DEBUG > /dev/null 2>&1 | cut -d ' ' -f 3-) <(./tools/test-documentation -L DEBUG > /dev/null 2>&1 | cut -d ' ' -f 3-)`
This command runs both scripts with `-L DEBUG`, ignores the first two "words" (time stamp) of each line of error output, and compares the error outputs of both scripts. 
We verified that `test-documentation-py-shash` was able to check for `zulip-py3-venv`.

timabbott commented:
@Shashantao can you clean up the commit history?

Check out the Zulip commit guidelines for more details: https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages

Shashantao commented:
@timabbott Does the commit history look better now?

timabbott commented:
Yes!  @Riken-Shah can you review this?

Riken-Shah commented on tools/test-documentation:
```diff
@@ -1,92 +1,104 @@
 # collapse_navigation is set to False in conf.py to improve sidebar navigation for users.
 # However, we must change its value to True before we begin testing links.
 # Otherwise, sphinx would generate a large number of links we don't need to test.
 # The crawler would take a very long time to finish and TravisCI would fail as a result.
```
Now we use Github Actions instead of TravisCI.

Shashantao replied:
Does that mean I need to remove Lines 56-63?

Riken-Shah replied:
No, I guess replace TravisCI on line 59 with Github Actions.

Shashantao replied:
I just did that.

Riken-Shah commented:
> We transpiled test-documentation to a Python script named test-documentation-py-shash. We compared the outputs of test-documentation and test-documentation-py-shash with the command line arguments --skip-check-links, --skip-external-links, -L ERROR, and -L DEBUG, respectively; we verified that the outputs of both scripts were equivalent with the following command for

I can also verify that both the outputs were similar after checking with different arguments.

Riken-Shah commented:
One thing I noticed strange is both the script `tools/test-documentation` and `tools/test-documentation` showed no error when my
virtual environment was not active.
Here are the screenshot, 
![Screenshot 2021-05-05 160245](https://user-images.githubusercontent.com/63820270/117129453-51a29b00-adbc-11eb-8dbe-e19cee7454bb.jpg)
![Screenshot 2021-05-05 160120](https://user-images.githubusercontent.com/63820270/117129461-536c5e80-adbc-11eb-95c2-e81da8749c5b.jpg)

Shashantao commented:
> One thing I noticed strange is both the script `tools/test-documentation` and `tools/test-documentation` showed no error when my
> virtual environment was not active.
> Here are the screenshot,
> ![Screenshot 2021-05-05 160245](https://user-images.githubusercontent.com/63820270/117129453-51a29b00-adbc-11eb-8dbe-e19cee7454bb.jpg)
> ![Screenshot 2021-05-05 160120](https://user-images.githubusercontent.com/63820270/117129461-536c5e80-adbc-11eb-95c2-e81da8749c5b.jpg)

@Riken-Shah I could not reproduce the problem in my WSL terminal. This probably had something to do with VS Code. My guess is that VS Code was itself running in the virtual env.
![屏幕截图 2021-05-05 193913](https://user-images.githubusercontent.com/74227626/117136103-7d347e00-adda-11eb-93f4-31e6fa2b7824.png)
![屏幕截图 2021-05-05 194031](https://user-images.githubusercontent.com/74227626/117136106-7e65ab00-adda-11eb-9fee-6cfae55bbde1.png)
![屏幕截图 2021-05-05 194206](https://user-images.githubusercontent.com/74227626/117136109-7efe4180-adda-11eb-804c-45530ef2694d.png)
![屏幕截图 2021-05-05 194216](https://user-images.githubusercontent.com/74227626/117136111-7efe4180-adda-11eb-9b39-df3b540cb1e3.png)

Riken-Shah commented:
> This probably had something to do with VS Code.

Yeah, I guess,
```console
riken@zulip:~/zulip$ which python3
/srv/zulip-py3-venv/bin/python3
riken@zulip:~/zulip$ 
```

Riken-Shah commented:
Yep! Everything LGTM!

@Shashantao Nothing major, but it'd be better if both commit messages had `test-documentation` as the prefix.
e.g: `test-documentation: Apply check for virtual environment.`
Also for the second commit, it will be more clear.
 `test-documentation: Replace outdated TravisCI comment with GitHub Actions.`

Shashantao commented:
@Riken-Shah I changed nothing in the code, but GitHub Actions just failed. A friend of mine had a similar experience in April:
https://chat.zulip.org/#narrow/stream/3-backend/topic/development.20sample.20messages/near/1167105
Probably I need to have GitHub Actions rerun, and I can't rerun it from my side. Any help?

Riken-Shah commented:
You can run the following commands to re-run the workflow,
```console
git commit --amend --no-edit
git push -f
```

Shashantao commented:
@Riken-Shah Things should be good now.

Riken-Shah commented:
One more thing, In https://github.com/zulip/zulip/pull/18124/commits/6645d9499bb6e7d9e278fd12bfe14fa30ca77e1c you don't need to add `Fixes #17929.`.

Shashantao commented:
> One more thing, In [6645d94](https://github.com/zulip/zulip/commit/6645d9499bb6e7d9e278fd12bfe14fa30ca77e1c) you don't need to add `Fixes #17929.`.

@Riken-Shah I moved `Fixes #17929.` from 957bf6e054e3be3c8c0bb6a51faa00b2b6333828 to [6645d94](https://github.com/zulip/zulip/commit/6645d9499bb6e7d9e278fd12bfe14fa30ca77e1c) because the latter was the final commit.

Riken-Shah commented:
>@Riken-Shah I moved Fixes #17929. from 957bf6e to 6645d94 because the latter was the last commit. Was Fixes #17929. supposed to be in the first or last commit?

The order doesn't matter! It should be in the commit which actually fixes it.

In your case, it's the first commit.

Shashantao commented:
@Riken-Shah Done.

Shashantao commented:
> > @Riken-Shah I moved Fixes #17929. from [957bf6e](https://github.com/zulip/zulip/commit/957bf6e054e3be3c8c0bb6a51faa00b2b6333828) to [6645d94](https://github.com/zulip/zulip/commit/6645d9499bb6e7d9e278fd12bfe14fa30ca77e1c) because the latter was the last commit. Was Fixes #17929. supposed to be in the first or last commit?
> 
> The order doesn't matter! It should be in the commit which actually fixes it.
> 
> In your case, it's the first commit.

@Riken-Shah I moved `Fixes #17929.` back to the first commit.

Riken-Shah commented:
GREAT @Shashantao!
Everything LGTM!

timabbott commented:
Closing in favor of https://github.com/zulip/zulip/pull/29325, thanks for working on this @Shashantao!
